File: Tests/Review6/bfs-dfs/bfs-dfs.py
from typing import List


class TestNode:

    def __init__(self, value: int = 0, neighbors: List = []):
        self.value: int = value
        self.neighbors: List[TestNode] = neighbors


def BFS(node: TestNode):
    queue = [node]
    layer = 0

    while len(queue):
        size = len(queue)
        for _ in range(size):
            current = queue.pop(0)
            if current:
                pass
            if len(node.neighbors) > 0:
                for item in node.neighbors:
                    queue.append(item)
        layer += 1


def DFS_Stack(root: TestNode):
    if not root:
        return

    stack: List[TestNode] = [root]

    while len(stack) > 0:
        current = stack.pop()
        print(current.value)
        for neighbor in current.neighbors:
            stack.append(neighbor)


def DFS_Recursive(root: TestNode):
    if not root:
        return

    def recursion(node: TestNode):
        if not node:
            return
        print(node.value)
        for neighbor in node.neighbors:
            recursion(neighbor)

    recursion(root)

File: Other/Review/Phase 8/Review4/Part1/bfs-dfs/bfs-dfs.ts
interface TestNode {
  value: any;
  neighbors: TestNode[];
}

export function BFS(root: TestNode): number {
  if (!root) {
    return -1;
  }

  const queue: TestNode[] = [root];
  let height: number = 0;

  while (queue.length) {
    const size = queue.length;
    for (let i = 0; i < size; i++) {
      const current = queue.shift()!;
      console.log(current.value);
      for (const neighbor of current.neighbors) {
        queue.push(neighbor);
      }
    }
    height++;
  }

  return height;
}

export function DFS_Iteration(root: TestNode) {
  if (!root) {
    return;
  }

  const stack: TestNode[] = [root];

  while (stack.length) {
    const current = stack.pop()!;
    console.log(current.value);
    for (const neighbor of current.neighbors) {
      stack.push(neighbor);
    }
  }
}

export function DFS_Recursion(root: TestNode) {
  if (!root) {
    return;
  }

  const recursion = (node: TestNode) => {
    if (!node) {
      return;
    }
    console.log(node.value);
    for (const neighbor of node.neighbors) {
      recursion(neighbor);
    }
  };

  recursion(root);
}

// test
const root: TestNode = {
  value: 1,
  neighbors: [
    {
      value: 2,
      neighbors: [
        {
          value: 4,
          neighbors: [],
        },
        {
          value: 5,
          neighbors: [],
        },
      ],
    },
    {
      value: 3,
      neighbors: [
        {
          value: 6,
          neighbors: [
            {
              value: 7,
              neighbors: [],
            },
          ],
        },
      ],
    },
  ],
};

// BFS
const res = BFS(root);
console.log('layer', res);

// DFS Recursion
console.log('------');
DFS_Recursion(root);

// DFS Iteration
console.log('------');
DFS_Iteration(root);
